system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,452 @@
+                }
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns default route on primary network service.
+    pub fn router(&self) -> Option<IpAddr> {
+        if let Some(router_str) = SCNetworkGlobal::query("ng_interface_router", "Router") {
+            if let Ok(router_ip) = router_str.parse::<IpAddr>() {
+                return Some(router_ip);
+            }
+        }
+
+        return None;
+    }
+
+    // pub fn netinfo(&self);
+    // pub fn proxies(&self) ;
+}
+
+/// Network service object.
+pub struct SCNetworkService(pub SCNetworkServiceRef);
+
+impl SCNetworkService {
+    /// Returns all available network services for the specified preferences.
+    pub fn list() -> Vec<SCNetworkService> {
+        let prefs = unsafe {
+            SCPreferencesCreate(
+                kCFAllocatorDefault,
+                CFString::from_static_string("ns_list").as_concrete_TypeRef(),
+                ptr::null(),
+            )
+        };
+
+        let array: CFArray<SCNetworkServiceRef> =
+            unsafe { CFArray::wrap_under_get_rule(SCNetworkServiceCopyAll(prefs)) };
+
+        array
+            .get_all_values()
+            .iter()
+            .map(|service_ptr| SCNetworkService(*service_ptr as _))
+            .collect::<Vec<SCNetworkService>>()
+    }
+
+    /// Returns the user-specified ordering of network services within the specified set.
+    pub fn list_order() -> Vec<SCNetworkService> {
+        let prefs = unsafe {
+            SCPreferencesCreate(
+                kCFAllocatorDefault,
+                CFString::from_static_string("ns_list_order").as_concrete_TypeRef(),
+                ptr::null(),
+            )
+        };
+
+        let netset = unsafe { SCNetworkSetCopyCurrent(prefs) };
+
+        let array: CFArray<SCNetworkServiceRef> =
+            unsafe { CFArray::wrap_under_get_rule(SCNetworkSetGetServiceOrder(netset)) };
+
+        let mut services = Vec::new();
+
+        for id in array.get_all_values().iter() {
+            let id_ptr: CFStringRef = *id as _;
+            let service_ptr: SCNetworkServiceRef = unsafe { SCNetworkServiceCopy(prefs, id_ptr) };
+            services.push(SCNetworkService(service_ptr));
+        }
+
+        services
+    }
+
+    /// Returns the identifier for this network service.
+    pub fn id(&self) -> String {
+        unsafe { CFString::wrap_under_get_rule(SCNetworkServiceGetServiceID(self.0)) }.to_string()
+    }
+
+    /// Returns the user-specified name associated with this network service.
+    pub fn name(&self) -> String {
+        unsafe { CFString::wrap_under_get_rule(SCNetworkServiceGetName(self.0)) }.to_string()
+    }
+
+    /// Returns this network service is enabled or disabled.
+    pub fn enabled(&self) -> bool {
+        let ret = unsafe { SCNetworkServiceGetEnabled(self.0) };
+        ret == 1
+    }
+
+    /// Returns the DNS infomation on this network service
+    pub fn dns(&self) -> SCNetworkServiceDNS {
```

Comment:
I'm a bit ambivalent about this method and `set_dns`. They are nice helpers indeed, but they are not at all an abstraction of the `SCNetworkConfiguration` API. They use `self` solely to get the network service ID, they don't call a single function in `system_configuration_sys::network_configuration::`.